correct height of cutouts within a layer (fix #8)

rdmueller opened the pull request:
This small change should end the cutouts for the top layer at the right height.

see #8 for example rendering.

Rendering looks good. Will print a test tomorrow.

Alex9779 commented on IKEA_Samla_Inserts.scad:
```diff
@@ -38,6 +38,11 @@ Layer_Marking_Type = "engrave"; // ["engrave":engrave, "emboss":emboss]
+// Cutouts
+// full cutouts are easier to print (no support needed)
+// correct cutouts let the top layer rest on the handle-cutouts
+Cutouts = "full"; //["full": full, "correct", correct]
```
This is placed in the "Layer marking" section, should be moved up before line 28.
Also this gives me three options in the "Customizer": "full", "correct" and "correct", last comma should be a semi-colon?

rdmueller replied:
ups. typo

rdmueller replied:
fixed this

Alex9779 commented:
Ok got a minute to have a look.
I am also not really happy with the naming, "correct" seems there was something wrong but this was a decision I made. Initially because for this to work right the cutout height has to be nearly correct which it wasn't and then I implemented the removed cutout if the top layer was smaller. And as told in the issue I didn't think about and consider this a valid special use case.
I would like to see it named like "Top Layer Cutouts", must not be the variable name, and the options "full" and "reduced" maybe (?) or "cut through" and "cutout height limited"... No real idea... :P

Thanks!

rdmueller commented:
you are right - I was also a bit unsure if "correct" would be a good wording. I will brainstorm a bit about it and come up with a new wording...

rdmueller commented:
I now switched to the wording "full" and "partial". I think this refelct the situation quite well.
I also renamed everything to TopLayerCutouts as you suggested.

Alex9779 commented on IKEA_Samla_Inserts.scad:
```diff
@@ -25,6 +25,11 @@ Addtional_Spacing = 1;
+// TopLayerCutouts
+// full TopLayerCutouts are easier to print (no support needed)
+// partial cutouts let the top layer rest on the handle-cutouts
+TopLayerCutouts = "full"; //["full": full, "partial": partial]
```
The first line should have spaces like the others.
Variable name should be "Top_Layer_Cutouts", underlines are rendered as spaces in the Customizer.
Also the sub-text seems too long, Customizer can't handle two lines:
![image](https://user-images.githubusercontent.com/6394581/120166451-df777780-c1fc-11eb-9edb-c883506d5b25.png)
Just the last one is displayed...

Suggestion:
```
// Top Layer Cutouts
// full are easier to print (no support needed), partial cutouts let the top layer rest on the handle-cutouts
Top_Layer_Cutouts = "full"; //["full": full, "partial": partial]
````

rdmueller replied:
hm. ok. But does the cutomizer work at all? You are speaking of the thingiverse customizer, aren't you? I didn't succeed with it.
Nevertheless, your suggestion is fine. I will fix this.

Alex9779 replied:
I actually meant the Customizer in OpenSCAD...

Alex9779 reviewed (changes requested):
Sorry but it seems this simple approach won't work out.
Switching to partial also shrinks the cutouts a little bit:
![Comparison](https://user-images.githubusercontent.com/6394581/120167448-0d10f080-c1fe-11eb-89ad-766ac17097ec.gif)
Hard to see in the animation but it is there.
The problem is the method I use to get the overall shrinkage of the cutouts to the top (also the widening of the box): scaling.
And scaling is calculated over the extrusion length (in this case the height) and a factor (here two factor one for x and one for y).
If you reduce the height and do not modify the factors the shape changes.
I remeber I had that problem initially when trying to just shrink the whole container to get the wall and bottom in one step. I changed then the bottom to be of a different thickness than the walls and so I had to generate it a different way and that problem was gone.
So the cutouts must be extruded the whole height, otherwise you have to adjust the scale factors which I don't think is easy. You could do some math to reduce them.
But I think an approach cutting of the top and regenerate without cutouts would be best.

Suggestion:
```
module Samla_Content(layer, width, depth, height, scale_width, scale_depth, width_handle, depth_handle, width_cutout, depth_cutout, scale_handle, scale_cutout, handle_cutout_height, diameter, diameter2, offset) {
    union() {
        difference() {
            linear_extrude(height=height, scale = [scale_width, scale_depth]) {
                offset(delta=-Addtional_Spacing+offset) Samla_Base(layer, width, depth, height, diameter, width_cutout, scale_cutout);
            }
            Samla_HandleAndCutout(layer, width, depth, height, scale_width, scale_depth, width_handle, depth_handle, width_cutout, depth_cutout, scale_handle, scale_cutout, handle_cutout_height, diameter, diameter2, offset);
        }
        if (Top_Layer_Cutouts == "partial") {
            intersection() {
                translate([0, 0, height-(height-handle_cutout_height)/2+(offset!=0?Bottom_Thickness:0)]) cube([width*scale_width, depth*scale_depth, height-handle_cutout_height], true);
                linear_extrude(height=height, scale = [scale_width, scale_depth]) {
                        offset(delta=-Addtional_Spacing+offset) Samla_Base(layer, width, depth, height, diameter, width_cutout, scale_cutout);
                }
            }
        }
    }
}
```

rdmueller commented:
I feared something like that. 
I printed 3 of those top layers for 5L boxes and the handle-cutouts did fit quite nice. I only gave the normal cutouts a bit more depth.
What I also noticed with my solution is that (especially for the normal cutouts) there seems to be a slight gap between the vertical and orizontal plane. Nothing to worry about in my case but not perfect.

Alex9779 commented:
> I feared something like that.

Well test my suggestion, no other change needed except the variable definition at the top, with that the partial cutouts don't change compared to full (in any layer).

Alex9779 commented:
Basically I am just doing what I described, I create the content again without cutouts, then a box to just get the upper top, slight hack needed to get the Bottom_Thickness right...

Alex9779 commented:
Still have to revert what you did in Samla_HandleAndCutout....

rdmueller commented:
ok, I applied your suggestion and got a nice rendering. I also sliced it and it seems that there is no gap between horizontal and vertical plane anymore :-) Thanx for your help!

Alex9779 commented:
> Still have to revert what you did in Samla_HandleAndCutout....

rdmueller commented:
ups. missed that

Alex9779 commented:
Thanks for your contribution!
